examples/vision/basnet_segmentation.py

```diff
@@ -0,0 +1,451 @@
+)
+val_dataset = load_dataset(
+    val_paths[0], val_paths[1], IMAGE_SIZE, OUT_CLASSES, BATCH_SIZE, shuffle=False
+)
+
+print(f"Train Dataset: {train_dataset}")
+print(f"Validation Dataset: {val_dataset}")
+
+"""
+## Visualize Data
+"""
+
+
+def display(display_list):
+    title = ["Input Image", "True Mask", "Predicted Mask"]
+
+    for i in range(len(display_list)):
+        plt.subplot(1, len(display_list), i + 1)
+        plt.title(title[i])
+        plt.imshow(keras.utils.array_to_img(display_list[i]), cmap="gray")
+        plt.axis("off")
+    plt.show()
+
+
+for image, mask in val_dataset.take(1):
+    display([image[2], mask[2]])
+
+"""
+## Analyze Mask
+
+Lets print unique values of above displayed mask. You can see despite belonging to one class, it's
+intensity is changing between low(0) to high(255). This variation in intensity makes it hard for
+network to generate good segmentation map for **salient or camouflaged object segmentation**.
+Because of its Residual Refined Module (RMs), BASNet is good in generating highly accurate
+boundaries and fine structures.
+"""
+
+print(f"Unique values count: {len(np.unique((mask[2] * 255)))}")
+print("Unique values:")
+print(np.unique((mask[2] * 255)).astype(int))
+
+"""
+## Building the BASNet Model
+
+BASNet comprises of a predict-refine architecture and a hybrid loss. The predict-refine
+architecture consists of a densely supervised encoder-decoder network and a residual refinement
+module, which are respectively used to predict and refine a segmentation probability map.
+
+![](https://i.imgur.com/8jaZ2qs.png)
+"""
+
+
+def basic_block(x_in, filters, stride=1, down_sample=None, no_relu=False):
+    """Creates a residual(identity) block with two 3*3 convolutions."""
+    residual = x_in
+
+    x = layers.Conv2D(filters, (3, 3), strides=stride, padding="same", use_bias=False)(
+        x_in
+    )
+    x = layers.BatchNormalization()(x)
+    x = layers.Activation("relu")(x)
+
+    x = layers.Conv2D(filters, (3, 3), strides=(1, 1), padding="same", use_bias=False)(
+        x
+    )
+    x = layers.BatchNormalization()(x)
+
+    if down_sample is not None:
+        residual = down_sample
+
+    x = layers.Add()([x, residual])
+
+    if not no_relu:
+        x = layers.Activation("relu")(x)
+
+    return x
+
+
+def convolution_block(x_in, filters, dilation=1):
+    """Apply convolution + batch normalization + relu layer."""
+    x = layers.Conv2D(filters, (3, 3), padding="same", dilation_rate=dilation)(x_in)
+    x = layers.BatchNormalization()(x)
+    return layers.Activation("relu")(x)
+
+
+def segmentation_head(x_in, out_classes, final_size):
+    """Map each decoder stage output to model output classes."""
+    x = layers.Conv2D(out_classes, kernel_size=(3, 3), padding="same")(x_in)
+
+    if final_size is not None:
+        x = layers.Resizing(final_size[0], final_size[1])(x)
+
+    return x
+
+
+def get_resnet_block(_resnet, block_num):

```

Comment:
Why does this use a `Model`? Wouldn't it be better in the Functional API?

Reply:
We can extract middle layers from existing model so I tried functional model too. But since we have to skip some initial and middle layers, doing functional way it was reducing the input dimension (due to initial layers) which was against the current architecture, If you don't pass input to the start of a model then it will not create a link between model input and output graph. so that's why it is not used.